SW-2792 Initial round of basic services in FE (with example usage)

karthikbtf opened the pull request:
- http service, main objective here is to abstract away all axios functionality, gives us flexibility on which implementation to use
- added preferences service
- added cached user service
- added user service
- remover api wrappers for functions implemented in services
- updated calls to refer to services
- this will iterate/evolve over time

karthikbtf commented on src/services/axios/index.ts:
```diff
@@ -0,0 +1,54 @@
+import axios from 'axios';
```
this was copied over from 'src/api/index.ts' - this is still valid and may need some clean up later.
currently the various api import axios from various locations ... abstracting this under HttpService gives more consistency/control over the underlying implementation

karthikbtf commented on src/components/Home.tsx:
```diff
@@ -68,7 +68,7 @@ export default function Home(): JSX.Element {
     let cancel = false;
     const populateUser = async () => {
-      const response = await getUser();
+      const response = await UserService.getUser();
```
this is a terrible example since we should be using the user from the UserProvider here... but just to illustrate..
I didn't move the user service usage to UserProvider .. yet.. needs more work

karthikbtf commented on src/services/UserService.ts:
```diff
@@ -0,0 +1,39 @@
+};
+
+const UserService = {
+  getUser,
```
this is just to show usage of http service. User service needs more functionality... will be separate effort

charlie-tf commented on src/services/HttpService.ts:
```diff
@@ -0,0 +1,160 @@
+    return await handleRequest(axios.put(replace(url, request), entity, { params, headers }));
+  };
+
+  const put = async (request: PutRequest = {}): Promise<Response> => post(request);
```
Curious about the choice to wrap `axios.put` into the `post` request type and then have `put` refer to `post`.. instead of wrapping it in the `put` and having `post` refer to that instead

karthikbtf replied:
oops, good catch.. that entire function was put and i renamed it to post but missed the axios part.. will fix.

karthikbtf replied:
fwiw - i have yet to use these functions in downstream services

karthikbtf commented on src/services/HttpService.ts:
```diff
@@ -0,0 +1,160 @@
+/**
+ * Url replacements
+ */
+export type Replacements = Record<string, string>;
```
TIL .. we have been using types like `{ [key: string]: string}` when we can just use `Record<string, string>` ,, handy

karthikbtf replied:
useful doc https://www.typescriptlang.org/docs/handbook/utility-types.html

charlie-tf replied:
Nice!

constanzauanini replied:
I think we have some Record<> in some parts of the code

charlie-tf reviewed (approved):
This looks really great. Thanks for doing this!

karthikbtf commented:
> This looks really great. Thanks for doing this!

Thanks, I will update this PR to exercise each of get/put/post/delete..

karthikbtf commented:
@charlie-tf apologies for the much larger change.. i think i will pause and get this in after Thu release.. and continue iterating.
The latest updates are:
- added preferences service
- added cached user service
- added update user
- remover api wrappers for functions implemented in services
- updated calls to refer to services

Unfortunately some of these changes had to go hand-in-hand for unbroken functionality.

If you get a chance to look again, I'd appreciate it.

constanzauanini commented:
> @charlie-tf apologies for the much larger change.. i think i will pause and get this in after Thu release.. and continue iterating. The latest updates are:
> 
> * added preferences service
> * added cached user service
> * added update user
> * remover api wrappers for functions implemented in services
> * updated calls to refer to services
> 
> Unfortunately some of these changes had to go hand-in-hand for unbroken functionality.
> 
> If you get a chance to look again, I'd appreciate it.



> @charlie-tf apologies for the much larger change.. i think i will pause and get this in after Thu release.. and continue iterating. The latest updates are:
> 
> * added preferences service
> * added cached user service
> * added update user
> * remover api wrappers for functions implemented in services
> * updated calls to refer to services
> 
> Unfortunately some of these changes had to go hand-in-hand for unbroken functionality.
> 
> If you get a chance to look again, I'd appreciate it.

I just took a look to all these. Like this new structure!

karthikbtf commented:
> > @charlie-tf apologies for the much larger change.. i think i will pause and get this in after Thu release.. and continue iterating. The latest updates are:
> > 
> > * added preferences service
> > * added cached user service
> > * added update user
> > * remover api wrappers for functions implemented in services
> > * updated calls to refer to services
> > 
> > Unfortunately some of these changes had to go hand-in-hand for unbroken functionality.
> > If you get a chance to look again, I'd appreciate it.
> 
> > @charlie-tf apologies for the much larger change.. i think i will pause and get this in after Thu release.. and continue iterating. The latest updates are:
> > 
> > * added preferences service
> > * added cached user service
> > * added update user
> > * remover api wrappers for functions implemented in services
> > * updated calls to refer to services
> > 
> > Unfortunately some of these changes had to go hand-in-hand for unbroken functionality.
> > If you get a chance to look again, I'd appreciate it.
> 
> I just took a look to all these. Like this new structure!

Thanks @constanzauanini

karthikbtf commented:
I am happy where this is right now, won't add any more changes.
Will get this in after Thu release and continue in new PRs.
What's here is functional.
